Simlar/src/SMLRLinphoneHandler.m

```diff
@@ -402,28 +469,86 @@ - (void)audioSessionInterrupted:(NSNotification *const)notification
-        SMLRLogE(@"Error while setting audio session to '%@': %@", active ? @"active" : @"inactive", error);
-        return;
+    switch (_callStatus.enumValue) {
+        case SMLRCallStatusNone:
+        case SMLRCallStatusEnded:
+            return AVAudioSessionCategorySoloAmbient; // default
+        case SMLRCallStatusIncomingCall:
+        case SMLRCallStatusConnectingToServer:
+        case SMLRCallStatusWaitingForContact:
+        case SMLRCallStatusRemoteRinging:
+        case SMLRCallStatusEncrypting:
+        case SMLRCallStatusTalking:
+            return AVAudioSessionCategoryPlayAndRecord;
     }
+}
 
-    if (active) {
-        [[AVAudioSession sharedInstance] setCategory:AVAudioSessionCategoryPlayAndRecord withOptions:AVAudioSessionCategoryOptionDuckOthers error:&error];
-        if (error != nil) {
-            SMLRLogE(@"Error while setting category of audio session: %@", error);
+- (AVAudioSessionCategoryOptions)generateAudioSessionCategoryOptions
+{
+    switch (_callStatus.enumValue) {
+        case SMLRCallStatusNone:
+        case SMLRCallStatusEnded:
+            return 0;
+        case SMLRCallStatusIncomingCall:
+            return AVAudioSessionCategoryOptionDefaultToSpeaker|AVAudioSessionCategoryOptionAllowBluetooth;
+        case SMLRCallStatusConnectingToServer:
+        case SMLRCallStatusWaitingForContact:
+        case SMLRCallStatusRemoteRinging:
+        case SMLRCallStatusEncrypting:
+        case SMLRCallStatusTalking:
+            return AVAudioSessionCategoryOptionMixWithOthers|AVAudioSessionCategoryOptionDuckOthers|AVAudioSessionCategoryOptionAllowBluetooth;
+    }
+}
+
+- (void)updateAudioSession
+{
+    SMLRLogFunc;
+
+    [self setAsyncAudioSessionCategory:[self generateAudioSessionCategory]
+                               options:[self generateAudioSessionCategoryOptions]
+                                  mode:AVAudioSessionModeVoiceChat];
+}
+
+- (void)setAsyncAudioSessionCategory:(NSString *const)category options:(const AVAudioSessionCategoryOptions)options mode:(NSString *const)mode
+{
+    dispatch_async(_audioSessionQueue, ^{
+        AVAudioSession *const sharedAudioSession = [AVAudioSession sharedInstance];
+
+        if (sharedAudioSession == nil) {
+            SMLRLogE(@"no shared audio session")
             return;
         }
 
-        [[AVAudioSession sharedInstance] setMode:AVAudioSessionModeVoiceChat error:&error];
-        if (error != nil) {
-            SMLRLogE(@"Error while setting mode of audio session: %@", error);
-            return;
+        if ([category isEqualToString:[sharedAudioSession category]] && options == [sharedAudioSession categoryOptions]) {
+            SMLRLogI(@"skipped setting audio session category: currentOptions=%lu", (unsigned long)[sharedAudioSession categoryOptions]);
+        } else {
+            SMLRLogI(@"setting audio session category: newOptions=%lu currentOptions=%lu", (unsigned long)options,(unsigned long)[sharedAudioSession categoryOptions]);
```

Comment:
But I think we do use a whitespace after the ',' when separating parameters. So:
`SMLRLogI(@"setting audio session category: newOptions=%lu currentOptions=%lu", (unsigned long)options, (unsigned long)[sharedAudioSession categoryOptions]);`